File: src/main/scala/jk_5/nailed/server/utils/InventoryOtherPlayer.scala
package jk_5.nailed.server.utils

import net.minecraft.entity.player.{EntityPlayer, EntityPlayerMP}
import net.minecraft.inventory.InventoryBasic

class InventoryOtherPlayer(private val owner: EntityPlayerMP, private val viewer: EntityPlayerMP) extends InventoryBasic(owner.getCommandSenderName + "\'s inventory", false, owner.inventory.mainInventory.length) {

  private var allowUpdate = false

  override def openInventory(player: EntityPlayer){
    InvSeeTicker.register(this)
    allowUpdate = false
    for(i <- 0 until owner.inventory.mainInventory.length){
      this.setInventorySlotContents(i, this.owner.inventory.mainInventory(i))
    }
    allowUpdate = true
    super.openInventory(player)
  }

  override def closeInventory(player: EntityPlayer){
    InvSeeTicker.unregister(this)
    if(allowUpdate){
      for(i <- 0 until owner.inventory.mainInventory.length){
        owner.inventory.mainInventory(i) = getStackInSlot(i)
      }
    }
    this.markDirty()
    super.closeInventory(player)
  }

  override def markDirty(){
    super.markDirty()
    if(allowUpdate){
      for(i <- 0 until owner.inventory.mainInventory.length){
        this.owner.inventory.mainInventory(i) = this.getStackInSlot(i)
      }
    }
  }

  def tick(){
    allowUpdate = false
    for(i <- 0 until owner.inventory.mainInventory.length){
      setInventorySlotContents(i, this.owner.inventory.mainInventory(i))
    }
    allowUpdate = true
    markDirty()
  }

  def getOwner = owner
  def getInventoryName = owner.getCommandSenderName + "\'s inventory"
}

File: src/main/java/jk_5/nailed/server/utils/InventoryOtherPlayer.java
package jk_5.nailed.server.utils;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.inventory.InventoryBasic;

public class InventoryOtherPlayer extends InventoryBasic {

    private final EntityPlayerMP viewer;
    private final EntityPlayerMP owner;
    private boolean allowUpdate;

    public InventoryOtherPlayer(EntityPlayerMP owner, EntityPlayerMP viewer) {
        super(owner.getCommandSenderName() + "'s inventory", false, owner.inventory.mainInventory.length);
        this.viewer = viewer;
        this.owner = owner;
    }

    @Override
    public void openInventory(EntityPlayer player){
        InvSeeTicker.register(this);
        allowUpdate = false;
        for(int id = 0; id < owner.inventory.mainInventory.length; ++id){
            setInventorySlotContents(id, owner.inventory.mainInventory[id]);
        }
        allowUpdate = true;
        super.openInventory(player);
    }

    @Override
    public void closeInventory(EntityPlayer player){
        InvSeeTicker.unregister(this);
        if(allowUpdate){
            for(int id = 0; id < owner.inventory.mainInventory.length; ++id){
                owner.inventory.mainInventory[id] = getStackInSlot(id);
            }
        }
        markDirty();
        super.closeInventory(player);
    }

    @Override
    public void markDirty(){
        super.markDirty();
        if(allowUpdate){
            for(int id = 0; id < owner.inventory.mainInventory.length; ++id){
                owner.inventory.mainInventory[id] = getStackInSlot(id);
            }
        }
    }

    public void update(){
        allowUpdate = false;
        for(int id = 0; id < owner.inventory.mainInventory.length; ++id){
            setInventorySlotContents(id, owner.inventory.mainInventory[id]);
        }
        allowUpdate = true;
        markDirty();
    }

    public EntityPlayer getOwner(){
        return owner;
    }
}
